catalog/app/utils/json-schema/json-schema.ts

```diff
@@ -0,0 +1,221 @@
+import Ajv from 'ajv'
+import * as dateFns from 'date-fns'
+import * as R from 'ramda'
+
+type CompoundCondition = 'anyOf' | 'oneOf' | 'not' | 'allOf'
+
+export type JsonSchema = Partial<
+  {
+    $ref: string
+    const: string
+    dateformat: string
+    default: any
+    enum: $TSFixMe[]
+    format: string
+    items: JsonSchema
+    properties: Record<string, JsonSchema>
+    type: string | string[] | JsonSchema[]
+  } & Record<CompoundCondition, JsonSchema[]>
+>
+
+// NOTE: can't use R.prop directly, because R.prop type doesn't allow undefined
+const prop = (
+  property: keyof JsonSchema,
+  optSchema?: JsonSchema,
+): JsonSchema[keyof JsonSchema] => (optSchema ? R.prop(property, optSchema) : undefined)
```

Comment:
i think we can default `optSchema` to `{}` and then just return `optSchema[property]`, and maybe call it `propSafe` or `propPartial` idk